### DISCUSSION TITLE
Mirror PSGallery

### DISCUSSION THREAD

**Discussion by rousseauxy** [bagetter/BaGetter#192] [Q&A]
Hi all, i was wondering if anyone had a working setup, and if it's even possible, to mirror psgallery into BaGetter?
I'm setting up a QA environment to replace our current nuget.server setup that hosts our own ps modules and psgallery modules.
It would be great if i could get the mirroring part working aswell.

In my appsettings i've adjust these settings
```
"Mirror": {
    "Enabled": true,
    "PackageSource":"https://www.powershellgallery.com/api/v2"
  },
```

but i'm getting the following error when trying to install a module.

```
Install-Module PSWindowsUpdate -Repository BaGet -Scope CurrentUser
WARNING: Unable to resolve package source 'http://localhost:5000/v3/index.json'.
PackageManagement\Install-Package : No match was found for the specified search criteria and module name 'PSWindowsUpdate'. Try Get-PSRepository to see all available registered modu
le repositories.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\1.0.0.1\PSModule.psm1:1809 char:21
+ ...          $null = PackageManagement\Install-Package @PSBoundParameters
+                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power....InstallPackage:InstallPackage) [Install-Package], Exception
    + FullyQualifiedErrorId : NoMatchFoundForCriteria,Microsoft.PowerShell.PackageManagement.Cmdlets.InstallPackage
```

If it's not possible that's good aswell, doesn't work with our current setup either, but was wondering if it's possible.

**Comment by Regenhardt** [comment]
Looks like PSGallery doesn't have a v3 api like nuget. So even though the v2 api is the same as nuget (Bagetter mirror supports that upstream as "legacy"), the powershell client can't call BaGetter as it does not contain the v2 api. PowerShell Core (pwsh) might support v3 though. Previous discussion in https://github.com/loic-sharma/BaGet/issues/427 is still up-to-date as we haven't done anything towards v2 since the fork.

**Reply by ghost** [reply]
Thanks for the update, it's no big deal, would have been a nice to have :)

So far setting up bagetter and importing our current modules into bagetter seemed to go smooth.
Need to test out our workflow now to see if it can fully replace our current nuget.server setup.

The only "issue" we noticed with powershell modules is that Powershell functions and Cmdlets are seen as tags for some reason, and as you might see on psgallery those are listed seperate :)

But will make an issue for it in the repo as an enhancement

**Reply by rousseauxy** [reply]
What possibility (and amount of work) would there be to implement the fork that [Baget made of V2](https://github.com/loic-sharma/BaGet/pull/694) and make it an option within the appsettings.json to enable the v2 support?

This would make bagetter a perfect onpremise psgallery repo/feed that supports both v2 and v3, making it possible to us the new PSResourceGet and the still widely used PowerShellGet in powershell 5.1 that is default shipped and installed on windows servers and clients.

I understand if this might not be on your scope, but I thought it wouldn't hurt to ask :)

**Reply by Regenhardt** [reply]
I could imagine trying that, even though all signs point towards a whole new can of worms there. I would absolutely review and help with someone's PR though :)

**Reply by rousseauxy** [reply]
Thanks for the answer, we recently did a poc of another product that did have those features and moved that to production :)
